src/wayland/xdg_foreign/mod.rs

```diff
@@ -0,0 +1,405 @@
+    }
+}
+
+#[derive(Debug)]
+struct Import {
+    inner: zxdg_imported_v2::ZxdgImportedV2,
+    surface: WlSurface,
+    handle: String,
+    /// Child surfaces which have imported this surface as a parent.
+    children: Vec<WlSurface>,
+}
+
+impl Import {
+    fn remove_children(&self, shell: &ShellState) {
+        for child in &self.children {
+            if let Some(child) = shell.toplevel_surface(child) {
+                // Make sure we are still the parent of the child surface?
+                if child.is_parent(&self.surface) {
+                    // Make the imported surface no longer the parent of this surface.
+                    child.set_parent(None);
+                }
+            }
+        }
+    }
+}
+
+fn implement_exporter(
+    state: Arc<Mutex<XdgForeignState>>,
+    exporter: Main<zxdg_exporter_v2::ZxdgExporterV2>,
+    shell_state: Arc<Mutex<ShellState>>,
+) -> zxdg_exporter_v2::ZxdgExporterV2 {
+    let destructor_state = state.clone();
+    let destructor_shell = shell_state.clone();
+
+    exporter.quick_assign(move |_, request, _| {
+        exporter_implementation(state.clone(), request, shell_state.clone());
+    });
+
+    exporter.assign_destructor(Filter::new(
+        move |exporter: zxdg_exporter_v2::ZxdgExporterV2, _, _| {
+            let state = destructor_state.lock().unwrap();
+            let mut exports = state.exports.borrow_mut();
+            // Since remove shifts the contents of the vector, we track the amount of removals so
+            // the next element is removed at it's proper index.
+            let mut removals = 0;
+
+            // This would be a whole lot nicer if there were a retain_mut
+            for index in 0..exports.len() {
+                let export = &mut exports[index - removals];
+
+                // If the export is from the client this exporter is destroyed from, then remove it.
+                if export.inner.as_ref().same_client_as(exporter.as_ref()) {
+                    // Destroy all imports created from this export's handle.
+                    export.destroy_imports();
+
+                    for import in &export.imports {
+                        import.remove_children(&*destructor_shell.lock().unwrap())
+                    }
+
+                    exports.remove(index - removals);
+                    removals += 1;
+                }
+            }
+        },
+    ));
+
+    exporter.deref().clone()
+}
+
+fn exporter_implementation(
+    state: Arc<Mutex<XdgForeignState>>,
+    request: zxdg_exporter_v2::Request,
+    shell: Arc<Mutex<ShellState>>,
+) {
+    match request {
+        zxdg_exporter_v2::Request::Destroy => {
+            // all is handled by destructor.
+        }
+
+        zxdg_exporter_v2::Request::ExportToplevel { id, surface } => {
+            // Check the role of the surface, this must be xdg_toplevel.
+            let role = compositor::get_role(&surface);
+
+            if role != Some("xdg_toplevel") {
```

Comment:
Same here, the spec says (emphasis mine)

> Only xdg_toplevel *equivalent* surfaces may be exported

so we should allow `zxdg_toplevel` and `wl_shell_surface` as well, and some may be added in the future.